src/v/cluster/rm_stm.cc

```diff
@@ -1790,19 +1773,50 @@ ss::future<raft::stm_snapshot> rm_stm::do_take_local_snapshot(
       [start_offset](tx_range range) { return range.last >= start_offset; });
     _aborted_tx_state.aborted = std::move(aborted);
 
+    if (!expired_abort_indexes.empty()) {
+        vlog(
+          _ctx_log.debug,
+          "Removing {} expired aborted indexes with offsets smaller than {}",
+          expired_abort_indexes.size(),
+          start_offset);
+    }
+    // grab the write lock to protect the aborted transaction indexes integrity
+    auto write_lock = co_await _state_lock.hold_write_lock();
+    chunked_vector<abort_snapshot> aborted_snapshots;
+    chunked_vector<abort_index> aborted_indexes_to_add;
     if (_aborted_tx_state.aborted.size() > _abort_index_segment_size) {
-        // There is nested locking here. Snapshot is done under
-        // persisted_stm::_op_lock and we grab a write lock to stop all the
-        // writers to aborted state to have a consistent snpahot to offload.
-        // Ensure this order is not violated in other places to avoid a
-        // deadlock.
-        auto units = co_await _state_lock.hold_write_lock();
-        co_await offload_aborted_txns();
+        std::sort(
+          std::begin(_aborted_tx_state.aborted),
+          std::end(_aborted_tx_state.aborted),
+          [](const tx_range& a, const tx_range& b) {
+              return a.first < b.first;
+          });
+
+        model::offset first = model::offset::max();
+        model::offset last = model::offset::min();
+        fragmented_vector<tx::tx_range> aborted_ranges;
+        for (auto const& entry : _aborted_tx_state.aborted) {
+            first = std::min(first, entry.first);
+            last = std::max(last, entry.last);
+            aborted_ranges.push_back(entry);
+            if (aborted_ranges.size() >= _abort_index_segment_size) {
+                aborted_snapshots.push_back(abort_snapshot{
+                  .first = first,
+                  .last = last,
+                  .aborted = std::move(aborted_ranges)});
+                aborted_indexes_to_add.emplace_back(first, last);
+                first = model::offset::max();
+                last = model::offset::min();
+                aborted_ranges = {};
+            }
+        }
+
+        _aborted_tx_state.aborted = std::move(aborted_ranges);
```

Comment:
we can not update `aborted` ranges outside of the apply lock as this collection is modified elsewhere

Reply:
i moved the lock releasing after we persist the aborted ranges snapshot

Reply:
> i moved the lock releasing after we persist the aborted ranges snapshot

Hmm doesn't it defeat the whole purpose of the pr as we'll still be waiting for disk ops under the apply lock?

Reply:
Can we do something like:
1. Calculate the abort ranges that we want to snapshot under the apply lock, save `aborted` and and `abort_indexes` that we are going to persist in the STM snapshot in local variables.
2. Release the apply lock
3. Persist new abort indexes
4. Update `aborted` and and `abort_indexes` in the STM state based on the indexes that we just persisted. Note that this could be different than what we saved in the snapshot, but I think we can update them correctly based on the list of indexes that we persisted.